src/rdsn/src/common/backup.thrift

```diff
@@ -64,24 +64,28 @@ struct configuration_restore_request
     9:optional string   restore_path;
 }
 
+// meta -> replica
 struct backup_request
 {
-    1:dsn.gpid              pid;
-    2:policy_info           policy;
-    3:string                app_name;
-    4:i64                   backup_id;
+    1:dsn.gpid          pid;
+    2:string            app_name;
+    3:i64               backup_id;
+    4:backup_status     status;
+    5:string            backup_provider_type;
     // user specified backup_path.
-    5:optional string       backup_path;
+    6:optional string   backup_root_path;
 }
 
 struct backup_response
 {
     1:dsn.error_code    err;
     2:dsn.gpid          pid;
-    3:i32               progress;  // the progress of the cold_backup
-    4:string            policy_name;
-    5:i64               backup_id;
-    6:i64               checkpoint_total_size;
+    3:i64               backup_id;
+    4:backup_status     status;
+    5:optional dsn.error_code   checkpoint_err;
+    6:optional dsn.error_code   upload_err;
```

Comment:
why need two error_code? using one-error_code + hint is not ok?

Reply:
Good idea, meta server can distinguish error with its backup_status.

Reply:
Ok, what the different between  `1:dsn.error_code    err` and `5:optional dsn.error_code   checkpoint_upload_err;`?

Reply:
`err` is used for backup_request rpc, such as ERR_OK, ERR_INVALID_STATE. `checkpoint_upload_err` is used for backup checkpoint  and upload. If backup checkpoint failed, checkpoint_upload_err won't be error_ok.

Reply:
So I think that just `err` is still OK. In other rpc, I remember that we have always defined it using one `err`.

Reply:
No, error is not always OK. For example, current partition is not primary, backup should be failed, err is ERR_INVALID_STATE.

Reply:
It's doesn't matter. If all passed, you return `ERR_OK`, otherwise, return `ERR_XXX`, you don't need return the case : `err = ERR_OK`, but `checkpoint=ERR_CHECKPOINT_FAILED`, just return `ERR_CHECKPOINT_FAILED` is ok.

Reply:
> No, error is not always OK

I mean that just need one variable `err` is enough

Reply:
I got your point. You think checkpoint_upload_err is unecessary, all errors can be shown by err. I think it is not clear if we combine two different error into one. `checkpoint_upload_err` is only used to show if checkpoint or upload meet error during backup. `err` is used for common error for backup request.

Reply:
If your error definition is appropriate enough, for example, I don't think a single err `ERR_ CHEKCPOINT_ FAILED` has different from  two err `err = ERR_ OK but err_checnpoint=ERR_ CHEKCPOINT_ FAILED`.

Reply:
In addition, hint_ message can also explain the error code, so there is no need to distinguish `common` and `backup`

Reply:
I don't agree with that one error code should identify all error cases. `checkpoint_upload_err` is only valid in backup checkpoint and upload stage, other stage it should be set. `err=ok` and `checkpoint_upload_err=ok` means backup not meet error during checkpoint and upload. It is clear to have two errors.
Besides, this rpc is sent from meta server and replica server periodically during backup, meta server will do different actions according to the err and checkpoint_upload_err, not only show the hint_message.
I still inisist on my previous design, seperate another error code.

Reply:
> meta server will do different actions

One err_code can't meet this requirement?

Reply:
In fact, only one error code can meet the requirement, but as I mentioned before I thought it was not clear enough, they are different error, should not be combined.

Reply:
I still think `checkpoint_upload_err` is redundant. if `ERR_CHECKPOINT_FAILED` can be clearly indicates that an error occurred in the checkpoint stage , why define `check_point_err` and set it value is `ERR_CHECKPOINT_FAILED` to explain?

In the past rpc, we haven't this design. Include bulkload, duplication, for example, we don't need add `ingest_err` to show the error of ingesting, or add `download_err` to show the err of downloading and so on, which is redundant, bu just use `ERR_INGEST_FAILD`, `ERR_DOWNLOAD` is clear enough.

Reply:
With your design, the sender needs to judge the response as follows:
```c++
if (rpc == ok) {
     if ( resp.err == ok) {
         if (resp.checkpoint_err == ok) {
             /** if you define `more clear err code as you say`, maybe need:**/
              /** if (err_a == ...) {
                        if (err_b == ...)
}
}
}
}
```
This logic is also redundant. don't suggest this design, you can refer the origin rpc define, I think the origin is elegant

Reply:
Actually, bulk load still seperete rpc_error, download_error and ingestion_err, you can reference sturcture `partition_bulk_load_state`.

And in my current design, won't have the too many if condition in your example. In my design, it just like
```
if (rpc != ok) { // handle this condition }
if (resp.err != ok) { // handle this condition }
if (resp.checkpoint_upload_err != ok) { // handle this condition }
```
If I combine error and checkpoint_upload_err into one, the code will be like:
```
if (rpc != ok) { // handle this condition }
if (resp.err != ok) { 
    if (original resp.err != ok){
        // handle this condition
    }
    if (original resp.checkpoint_upload_err != ok ) {
        // handle this condition
    }
}
```

I think it is okay to have two errors to distinguish different errors, if there is only one field redundant, but can make structure clear. In my previous design, checkpoint error and upload error also should be seperated, I have already compromised my logic :-)

Reply:
Well, wait other's opinion, if they have no have no objections, I will be willing think it is a reasonable design

Reply:
@acelyc111 Please give us your suggestion about this comment~ @foreverneverer has different opion with me, and we can not persude each other.

Reply:
If I didn't miss anything, I didn't find where `checkpoint_upload_err` is used currently. We keep this point in mind and see how will it be used later, if found it's not so elegant, we can point it out then.